[AspNetCore] PublishToFolder

jtorres opened the pull request:
Fixes https://devdiv.visualstudio.com/DevDiv/_workitems/edit/669822

# TODO

- [x] Rewrite serializer
- [x] Relative paths handling
- [x] Full VS for windows compatibility check
    - [x] pubxml files generated with VSforMac are comaptibles with VS
    - [x] pubxml files generated with VS are comaptibles with VSforMac (FAILING : WIP )
    - [x] WebPublishMethod diff that FileSystem are ~disabled~ _hidden_ on VSforMac
- [x] Minors TODO's
- [x] field validator in Dialog
- [x] Menu `Publish` locations is wrong?
- [x] ~Do we have to implement `DeleteExistingFiles`~
- [x] Close Dialog as soon as Publish is clicked
- [x] Call Save All (like Build) before Publish is clicked
- [x] Reuse if possible `Publish pad`
- [x] Telemetry
- [ ] Usability (a11y)

Commit:
[AspNetCore] Implements PublishToFolder

Commit:
[AspNetCore] Few fixes and improvements

Commit:
[AspNetCore] Better dotnetcore runtime detection check

Commit:
[AspNetCore] Splits PublishToFolder Command

Commit:
[AspNetCore] Sets extension point's path

Commit:
[AspNetCore] Implements pubxml file handling

Commit:
[AspNetCore] Makes publishUrl relative paths

Commit:
[AspNetCore] Fixes minor issues

- Adds current active Platform
- Fixes GetActiveConfig issue

Commit:
[AspNetCore] Sets Platform explicitly if it's empty

Commit:
[AspNetCore] Fixes a few compatibility issues: Windows - Mac

Commit:
[AspNetCore] Implements path field validator

rodrmoya commented on main/src/addins/MonoDevelop.AspNetCore/MonoDevelop.AspNetCore.Commands/PublishToFolderBaseCommandHandler.cs:
```diff
@@ -0,0 +1,153 @@
+						}
+					}
+					CloseDialog ();
+					progressMonitor.EndTask ();
```
Shouldn't these 2 lines be in a `finally` block, just in case something fails, the dialog should be closed always and the progressMonitor notified always also, right?

rodrmoya commented on main/src/addins/MonoDevelop.AspNetCore/MonoDevelop.AspNetCore.Commands/PublishToFolderBaseCommandHandler.cs:
```diff
@@ -0,0 +1,153 @@
+
+		protected static bool ProjectSupportsAzurePublishing (DotNetProject project)
+		{
+			return project != null && project.GetProjectCapabilities ().Any (i => i == "Web" || i == "AzureFunctions");
```
is this the correct way to check this? I think @mrward mentioned adding this to project capabilities, one that says the project is "publishable".

rodrmoya replied:
Also, the method is misnamed, as this is not about Azure publishing but about the publishing to a folder

jtorres replied:
regarding to "publishable" : I missed that... so we want to add that capability?

slluis replied:
Capabilities are provided by msbuild, we can't really add new capabilities. If .Net Core msbuild projects emit a "publishable" capability, we can use that. Otherwise, it is safer to just check the project type.

rodrmoya commented on main/src/addins/MonoDevelop.AspNetCore/MonoDevelop.AspNetCore.Commands/PublishToFolderProfilesCommandHandler.cs:
```diff
@@ -0,0 +1,32 @@
+			}
+
+			foreach (var profile in profiles.OrderBy (x => x.Name)) {
+				if (profile.WebPublishMethod == "FileSystem")
```
hmm, if "profiles" is null, this would crash, wouldn't it? I think this `foreach` should be inside the `if` above?

jtorres replied:
not really, `GetPublishProfiles()` does not return `null`.. but the `if (profiles != null)` statement is not necessary though

slluis replied:
The is no need to conditionally add the separator. The menu system ensures that if the separator is not needed, it won't be shown.

rodrmoya commented on main/src/addins/MonoDevelop.AspNetCore/MonoDevelop.AspNetCore.Dialogs/PublishToFolderDialog.cs:
```diff
@@ -0,0 +1,154 @@
+		Label MessageLabel;
+		ImageView MessageIcon;
+		Uri BinBaseUri => new Uri (Path.Combine (publishCommandItem.Project.BaseDirectory, "bin"));
+
```
just a styling nitpick, since all these are private fields, they shouldn't be using capital letters as first char

rodrmoya reviewed (commented):
Apart from these, code LGTM and works great!

Commit:
[AspNetCore] Fixes an issue when creating pubxml file

Commit:
[AspNetCore] WIP more fixes

- adds telemetry counter (WIP)
- modifies naming
- sets project level Publish command

jtorres committed:
[AspNetCore] Sorts out Menu

rodrmoya commented on main/src/core/MonoDevelop.Ide/ExtensionModel/MainMenu.addin.xml:
```diff
@@ -221,6 +221,8 @@
 		<CommandItem id = "MonoDevelop.Ide.Commands.ProjectCommands.Clean" />
 		<SeparatorItem id = "Separator1" />
 		<CommandItem id = "MonoDevelop.Ide.Commands.ProjectCommands.Stop" />
+		<Separator id = "Separator2">
```
```suggestion
		<Separator id = "Separator2" />
```

Therzok commented on main/src/addins/MonoDevelop.AspNetCore/MonoDevelop.AspNetCore.Commands/ProjectPublishProfile.cs:
```diff
@@ -0,0 +1,118 @@
+	}
+
+	[XmlRoot ("PropertyGroup", Namespace = "http://schemas.microsoft.com/developer/msbuild/2003")]
+	public sealed class ProjectPublishProfile : PublishProfile
```
This doesn't need to be public?

jtorres replied:
yes, looks like it must be public, otherwise it can't be serialized... turning it back to public

Therzok commented on main/src/addins/MonoDevelop.AspNetCore/MonoDevelop.AspNetCore.Commands/PublishCommandItem.cs:
```diff
@@ -0,0 +1,20 @@
+			this.Profile = profile;
+		}
+
+		public bool IsReentrant { get; private set; }
```
```suggestion
		public bool IsReentrant { get; set; }
```

Commit:
[Ide] Add missing closing XML tag

Therzok commented on main/src/addins/MonoDevelop.AspNetCore/MonoDevelop.AspNetCore.Commands/PublishCommandItem.cs:
```diff
@@ -0,0 +1,20 @@
+
+		public bool IsReentrant { get; private set; }
+
+		public DotNetProject Project { get; private set; }
```
```suggestion
		public DotNetProject Project { get; set; }
```

Therzok replied:
```suggestion
		public DotNetProject Project { get; }
```

Therzok commented on main/src/addins/MonoDevelop.AspNetCore/MonoDevelop.AspNetCore.Commands/PublishCommandItem.cs:
```diff
@@ -0,0 +1,20 @@
+
+		public DotNetProject Project { get; private set; }
+
+		public ProjectPublishProfile Profile { get; set; }
```
Shouldn't setting this change the Reentrant property?

jtorres replied:
it does not, Reentrant is set in the constructor so that when the instance is created, at this point we know, depending on `profile`, if this is Reentrant or not, by setting it in the `Profile` setter it brakes the expected behaviour

Therzok commented on main/src/addins/MonoDevelop.AspNetCore/MonoDevelop.AspNetCore.Commands/PublishToFolderBaseCommandHandler.cs:
```diff
@@ -0,0 +1,158 @@
+				if (!(dataItem is PublishCommandItem publishCommandItem)) {
+					dialog = new PublishToFolderDialog (new PublishCommandItem (project, null));
+					dialog.PublishToFolderRequested += Dialog_PublishToFolderRequested;
+					if (dialog.Run () == Xwt.Command.Close) {
```
```suggestion
					WindowFrame parent = Toolkit.CurrentEngine.WrapWindow (IdeApp.Workbench.RootWindow);
					if (dialog.Run (parent) == Xwt.Command.Close) {
```

Because we need it to be modal.

Therzok commented on main/src/addins/MonoDevelop.AspNetCore/MonoDevelop.AspNetCore.Dialogs/PublishToFolderDialog.cs:
```diff
@@ -0,0 +1,154 @@
+				Name = "BrowseEntryHBox",
+				Spacing = 4
+			};
+			var defaultDirectory = publishCommandItem.Project.GetActiveConfiguration () == null
```
```suggestion
			var activeConfig = publishCommandItem.Project.GetActiveConfiguration ();
			var defaultDirectory = activeConfiguration == null
			? BinBaseUri.ToString ()
				: Path.Combine (BinBaseUri.ToString (),
								publishCommandItem.Project.TargetFramework.Id.GetShortFrameworkName (),
								activeConfiguration);
```

Commit:
[Ide] Fix incorrect XML tag

Therzok commented on main/src/addins/MonoDevelop.AspNetCore/MonoDevelop.AspNetCore.Dialogs/PublishToFolderDialog.cs:
```diff
@@ -0,0 +1,154 @@
+				Name = "PathEntry",
+				Text = defaultDirectory
+			};
+			PathEntry.Changed += PathEntry_Changed;
```
Events not unsubscribed.

rodrmoya commented on main/src/addins/MonoDevelop.AspNetCore/MonoDevelop.AspNetCore/Counters.cs:
```diff
@@ -0,0 +1,41 @@
+	internal class BuildOutputCounterMetadata : CounterMetadata
+	{
+		public MSBuildVerbosity Verbosity {
+			get => GetProperty<MSBuildVerbosity> ("Verbosity");
```
All these are a copy paste from the build output, so should be removed or replace with appropriate properties for PublishToFolder telemetry

jtorres replied:
telemetry is WIP

jtorres committed:
[AspNetCore] Removes WIP telemetry implementation

jtorres committed:
[AspNetCore] Renames fields

jtorres committed:
[AspNetCore] Few fixes

- Removes redundant checks
- Improves interpolation calls
. Makes dialog modal
- Sorts out accesors

jtorres committed:
[AspNetCore] Sets back ProjectPublishProfile to public

jtorres committed:
[AspNetCore] Close Dialog after clicks Publish

jtorres committed:
[AspNetCore] Ensures to SaveAll before publishing

Commit:
[AspNetCore] Add telemetry counter for Publish to Folder

Commit:
[AspNetCore] Add basic ProjectPublishProfile tests

Commit:
[AspNetCore] Output ASP.NET Core tests so that they can be found in VSTS

jtorres committed:
[AspNetCore] Sets a standard progressmonitor name so that it could be reused

Therzok commented on main/src/addins/MonoDevelop.AspNetCore/MonoDevelop.AspNetCore.Tests/MonoDevelop.AspNetCore.Tests/PublishProfileTests.cs:
```diff
@@ -0,0 +1,59 @@
+
+			var readProfile = ProjectPublishProfile.ReadModel (tmpFile);
+
+			Assert.That (profile.LastUsedBuildConfiguration, Is.EqualTo (readProfile.LastUsedBuildConfiguration));
```
❓ Maybe it would be useful to make `ProjectPublishProfile` implement `IEquatable<ProjectPublishProfile>` and just test that they're equal.

Therzok replied:
That way the test does not go stale once more properties are added

jtorres committed:
[AspNetCore] Avoids ref cycle and unsubscribes event handler

jtorres committed:
[AspNetCore] Sets default folder to Release

Therzok commented on main/Main.sln:
```diff
@@ -261,15 +261,15 @@ Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "FSharpBinding", "FSharpBind
 	EndProjectSection
 EndProject
-Project("{F2A71F9B-5D33-465A-A702-920D77279786}") = "MonoDevelop.FSharp", "external\fsharpbinding\MonoDevelop.FSharpBinding\MonoDevelop.FSharp.fsproj", "{4C10F8F9-3816-4647-BA6E-85F5DE39883A}"
+Project("{f2a71f9b-5d33-465a-a702-920d77279786}") = "MonoDevelop.FSharp", "external\fsharpbinding\MonoDevelop.FSharpBinding\MonoDevelop.FSharp.fsproj", "{4C10F8F9-3816-4647-BA6E-85F5DE39883A}"
```
Diff noise here.

Therzok commented on main/Main.sln:
```diff
@@ -349,6 +351,8 @@ Global
 		ReleaseGnome|Any CPU = ReleaseGnome|Any CPU
 		ReleaseMac|Any CPU = ReleaseMac|Any CPU
 		ReleaseWin32|Any CPU = ReleaseWin32|Any CPU
+		Debug|x86 = Debug|x86
```
This shouldn't be written.

Commit:
Revert unwanted changes in .sln file

rodrmoya reviewed (approved):
Apart from that leftover commented code, this LGTM and works ok, so is ready to be merged IMO (tests will be done separately, as they need a bit more work to reuse some of the .NET Core test "infrastructure")

Therzok commented on main/src/addins/MonoDevelop.AspNetCore/MonoDevelop.AspNetCore.Commands/PublishToFolderBaseCommandHandler.cs:
```diff
@@ -0,0 +1,169 @@
+				using (Counters.PublishToFolder.BeginTiming (counterMetadata)) {
+					try {
+
+						IdeApp.Workbench.SaveAll ();
```
Why do we do this? Nothing else in the IDE saves before building?

Therzok replied:
If you want to prompt for save before publishing, call [ApplySavePolicy](http://source.monodevelop.com/#MonoDevelop.Ide/MonoDevelop.Ide/ProjectOperations.cs,cef4c1c03192c28b,references)

rodrmoya commented:
@monojenkins backport release-7.7

rodrmoya commented:
BTW, md-addins related branch -> https://github.com/xamarin/md-addins/pull/3739

monojenkins commented:
@rodrmoya backporting to release-7.7 failed, the patch results in conflicts:

```
Applying: [AspNetCore] Implements PublishToFolder
Applying: [AspNetCore] Few fixes and improvements
Applying: [AspNetCore] Better dotnetcore runtime detection check
Applying: [AspNetCore] Splits PublishToFolder Command
Applying: [AspNetCore] Sets extension point's path
Applying: [AspNetCore] Implements pubxml file handling
Applying: [AspNetCore] Makes publishUrl relative paths
Applying: [AspNetCore] Fixes minor issues
Applying: [AspNetCore] Sets Platform explicitly if it's empty
Applying: [AspNetCore] Fixes a few compatibility issues: Windows - Mac
Applying: [AspNetCore] Implements path field validator
Applying: [AspNetCore] Fixes an issue when creating pubxml file
Applying: [AspNetCore] WIP more fixes
Applying: [AspNetCore] Sorts out Menu
Applying: [Ide] Add missing closing XML tag
Applying: [Ide] Fix incorrect XML tag
Applying: [AspNetCore] Removes WIP telemetry implementation
Applying: [AspNetCore] Renames fields
Applying: [AspNetCore] Few fixes
Applying: [AspNetCore] Sets back ProjectPublishProfile to public
Applying: [AspNetCore] Close Dialog after clicks Publish
Applying: [AspNetCore] Ensures to SaveAll before publishing
Applying: [AspNetCore] Add telemetry counter for Publish to Folder
Applying: [AspNetCore] Add basic ProjectPublishProfile tests
Using index info to reconstruct a base tree...
M	main/Main.sln
M	main/src/addins/AspNet/MonoDevelop.AspNet.csproj
M	main/src/addins/MonoDevelop.AspNetCore/MonoDevelop.AspNetCore/ProjectExtensions.cs
.git/rebase-apply/patch:306: trailing whitespace.
// Copyright (c) 2018 
.git/rebase-apply/patch:328: trailing whitespace.
// Information about this assembly is defined by the following attributes. 
.git/rebase-apply/patch:346: trailing whitespace.
// The following attributes are used to specify the signing key for the assembly, 
warning: 3 lines add whitespace errors.
Falling back to patching base and 3-way merge...
Auto-merging main/src/addins/MonoDevelop.AspNetCore/MonoDevelop.AspNetCore/ProjectExtensions.cs
CONFLICT (content): Merge conflict in main/src/addins/MonoDevelop.AspNetCore/MonoDevelop.AspNetCore/ProjectExtensions.cs
Auto-merging main/src/addins/AspNet/MonoDevelop.AspNet.csproj
Auto-merging main/Main.sln
error: Failed to merge in the changes.
Patch failed at 0024 [AspNetCore] Add basic ProjectPublishProfile tests
```

Please backport manually!
